base/shell/explorer/explorer.cpp

```diff
@@ -87,6 +87,56 @@ HideMinimizedWindows(IN BOOL bHide)
 }
 #endif
 
+static BOOL
+IsExplorerSystemShell()
+{
+    HKEY hKeyWinlogon;
+    WCHAR szPath[MAX_PATH];
+    LPWSTR szExplorer;
+    BOOL bIsSystemShell = TRUE; // Assume we are the system shell by default.
+
+    if (!GetModuleFileNameW(NULL, szPath, _countof(szPath)))
+        return FALSE;
+
+    szExplorer = PathFindFileNameW(szPath);
+
+    if (RegOpenKeyExW(HKEY_LOCAL_MACHINE, L"Software\\Microsoft\\Windows NT\\CurrentVersion\\Winlogon",
+        0, KEY_READ, &hKeyWinlogon) != ERROR_SUCCESS)
+    {
+        // No registry access.
+        bIsSystemShell = TRUE;
+    }
+    else
+    {
+        DWORD cbShell;
+        LSTATUS Status;
+
+        // Get length of the "Shell" key
+        Status = RegQueryValueExW(hKeyWinlogon, L"Shell", 0, NULL, NULL, &cbShell);
```

Comment:
As the code currently stands, it would be more logical to do this "get length" only when someone actually implements the dynamical path allocation.

Reply:
He means only one call to RegQueryValueExW

Reply:
Latest commit brings it down to one call to RegQueryValueExW

Reply:
`
        LSTATUS Status;
        DWORD dwType;
        WCHAR szShell[MAX_PATH];
        DWORD cbShell = sizeof(szShell);

        // TODO: Add support for paths longer than MAX_PATH
        Status = RegQueryValueExW(hKeyWinlogon, L"Shell", 0, &dwType, (LPBYTE)szShell, &cbShell);
        if (Status == ERROR_SUCCESS)
        {
            if ((dwType == REG_SZ || dwType == REG_EXPAND_SZ) && StrStrI(szShell, szExplorer))
                bIsSystemShell = TRUE;
            else
                bIsSystemShell = FALSE;
        }

        RegCloseKey(hKeyWinlogon);`